### ISSUE TITLE
Slow startup for v3

### ISSUE THREAD

**Issue by daniporr** [desktop/desktop#14513]
### The problem

Version 3.0.0 takes much more time (like 3x, but I cannot measure it exactly) to start than previous versions (on the same machine).

### Release version

3.0.0

### Operating system

Windows 10

### Steps to reproduce the behavior

1. update to version 3.0.0
2. start GitHub Desktop

### Additional context

It would be interesting to know if someone else experienced the same issue (otherwise it might be something in my machine)

**Comment by sergiou87** [comment]
Thank you for your report. I don't think there is any special reason for 3.0.0 to take longer to launch compared to previous versions 🤔 There is nothing major I can think of that could affect that.

Do you use an antivirus or something that could slow it down?

**Comment by daniporr** [comment]
Thank you for you help.

Yes, I use the default antivirus in Windows (but I was using it also before).

Since you mentioned that, I have tried to disable it and run GitHub Desktop, but nothing changed.

Is there anything else I can try?
(since nobody voted on this issue, I am starting to believe that it is just a problem on my machine)

**Comment by sergiou87** [comment]
@daniporr I can only think of a complete reinstall by means of

1. Quitting GitHub Desktop and restarting your machine to ensure no rogue processes are hanging about
1. Uninstalling GitHub Desktop
1. Deleting (or moving/renaming temporarily) the `C:\Users<YourUserName>\AppData\Roaming\GitHub Desktop` folder. This folder has all your GitHub Desktop settings and the list of repositories, so you might want to keep a copy of it.
1. Downloading and installing GitHub Desktop again from https://desktop.github.com

**Comment by daniporr** [comment]
@sergiou87 I tried that, but it did not work.

I have also waited a new version (to try that with a new version), but the latest one is still 3.0.0 (although it is now more than one month old).

However, it seems to just be a problem of my machine for the moment, so I think I just need to come to terms with that...

Thank you for your help.

**Comment by daniporr** [comment]
Just to provide more info on this, from the log I noticed that there are some operations done at the beginning (before starting the GUI) which take really a lot of time:

```log
info: [ui] Executing getRemotes: git remote -v (took 4.784s)
info: [ui] Executing getStatus: git --no-optional-locks status --untracked-files=all --branch --porcelain=2 -z (took 2.081s)
info: [ui] Executing getRecentBranches: git log -g --no-abbrev-commit --pretty=oneline HEAD -n 2500 -- (took 1.403s)
info: [ui] Executing installGlobalLFSFilter: git lfs install --skip-repo (took 3.306s)
```

I was wondering whether all of them are required when we start the application (in some cases I open it just to check previous commits that I have locally so these operations seem unnecessary).
I feel that some of them could be postponed until after clicking on a specific repository (and executed only if needed).

Moreover, probably the GUI should start anyway and then perform (if needed) these additional operations.

---

P.S. this is false

> We haven’t gotten a response to our questions above.

**Comment by daniporr** [comment]
Do you need further information about that?

**Comment by sergiou87** [comment]
Hey @daniporr ! Sorry I missed the notifications from your responses.

Those slow git calls shouldn't take that long unless maybe your repository is huge or it need some repacking with `git gc` (see https://git-scm.com/docs/git-gc)

I would also suggest you to try the latest beta (https://desktop.github.com/beta) which uses a newer version of Electron that might impact the performance of the app.

**Comment by daniporr** [comment]
@sergiou87 Thank you very much for your suggestions!
I will try them in the next few days.

**Comment by daniporr** [comment]
@sergiou87 unfortunately they do not seem to help.

Could the number of repositories loaded in GitHub Desktop affect the performance?

However, I am still wondering whether those operations are "necessary":

> I was wondering whether all of them are required when we start the application (in some cases I open it just to check previous commits that I have locally so these operations seem unnecessary).
> I feel that some of them could be postponed until after clicking on a specific repository (and executed only if needed).
> 
> Moreover, probably the GUI should start anyway and then perform (if needed) these additional operations.

Does it make sense to postpone them?

**Comment by sergiou87** [comment]
> Could the number of repositories loaded in GitHub Desktop affect the performance?

Hmmm it's possible Desktop makes some `git` calls to repos in your list to indicate, in the repo list, if there are new commits for every specific repo. You can, again, backup your settings (the `C:\Users<YourUserName>\AppData\Roaming\GitHub Desktop` folder), and start removing repositories from your list to see if that makes a difference.

I can also provide download links to previous versions of Desktop to see if that makes a difference, just let me know and I'll share some here 😄 

> Does it make sense to postpone them?

Maybe! I mean, there is always room for optimization, but we don't want to optimize things that shouldn't be slow to begin with, because that might have other costs 😢 

Could you run any/some of those commands manually in the same repo from your terminal and see if they take that long to complete?

**Comment by daniporr** [comment]
> Hmmm it's possible Desktop makes some `git` calls to repos in your list to indicate, in the repo list, if there are new commits for every specific repo. You can, again, backup your settings (the `C:\Users<YourUserName>\AppData\Roaming\GitHub Desktop` folder), and start removing repositories from your list to see if that makes a difference.

I tried to do this, and, unless I have no repositories in GitHub Desktop (which would make it quite useless), it still runs those commands at the beginning (and it requires a lot of time to run them).

> I can also provide download links to previous versions of Desktop to see if that makes a difference, just let me know and I'll share some here 😄

It was not there before (I have been using GitHub Desktop for a long time).
It started with v3 (or soon before that, maybe I have not updated all intermediate releases).

> Maybe! I mean, there is always room for optimization, but we don't want to optimize things that shouldn't be slow to begin with, because that might have other costs 😢

Probably I haven't explained clearly my point.

I think that tasks which are not strictly needed to draw the UI should not block it. You can run them in parallel, but if they are not required to start the GUI, they should not interfere with it.
In general (as you probably already know) an activity (in this case those git commands) should not block another one (in this case the GUI) if it is not essential for that one.
Probably nobody noticed this design error because those tasks are very fast on powerful machines (which I think most developers use), but they have not considered that users might have not very powerful machines.
Therefore, I think that here is not matter of optimizing, there is actually a design error (those processes should not be "blocking").

> Could you run any/some of those commands manually in the same repo from your terminal and see if they take that long to complete?

It depends on the repository, but they roughly take the same time reported in the logs (probably it is just because my machine is slow).
But again I think they should not be there if they are not needed.

Moreover, I think it is fine if you want to "precompute" something to speedup future operations, but those computations should not affect the performance.

---

Thank you again for your help.
Let me know if you need more information about that.

**Comment by sergiou87** [comment]
> I think that tasks which are not strictly needed to draw the UI should not block it. You can run them in parallel, but if they are not required to start the GUI, they should not interfere with it.
> In general (as you probably already know) an activity (in this case those git commands) should not block another one (in this case the GUI) if it is not essential for that one.

Oh! Yeah, I misunderstood what you said. In theory those operations always run asynchronously and not block the UI, if they do it's definitely a bug!! I'm 99% sure they are running in parallel (or that was our intention from the very beginning), but I'll schedule some time to double-check and make sure we're not blocking the UI, which would be pretty terrible 😱

**Comment by daniporr** [comment]
Thank you very much 😃

**Comment by ISAACMETHOD** [comment]
You welcome

**Comment by sergiou87** [comment]
@daniporr I just tried adding some artificial `sleep(10)` to delay those git tasks for extra 10 seconds, and I confirmed the app is running those completely asynchronously. It waits for them to complete before populating the UI with the right info, but until then the app stays functional with this empty UI:

![image](https://user-images.githubusercontent.com/1083228/187180807-30ddbcbd-0e67-42ab-aeaa-8b49610495f6.png)

So I think there must be something else going on 😕

**Comment by daniporr** [comment]
@sergiou87 thank you for checking that.

Those were the only statements in the log file related to the initial loading time.
I do not know what else I should test. Let me know if you have any other idea about that.

Thanks again for your help.

**Comment by daniporr** [comment]
I think we can close this issue now.
We can open new ones for newer versions if needed.
